src/odemis/gui/cont/fastem_acq.py

```diff
@@ -581,32 +563,231 @@ def _on_alignment_done(self, future):
+                futures[f] = t
+
+        # if futures:
+        calib_future = model.ProgressiveBatchFuture(futures)
+        # also handles cancelling and exceptions
+        calib_future.add_done_callback(self._on_batch_calibrations_done)
+        # connect the future to the progress bar and its label
+        self._future_connector = ProgressiveFutureConnector(calib_future, self.gauge, self.label, full=False)
+        # TODO needed?
+        # else:  # In case all calibrations failed to start
+        #     self._main_data_model.is_acquiring.value = False
+        #     self._reset_calibration_gui("Calibrations failed (see log panel).", level=logging.WARNING)
+
+    # @call_in_wx_main
+    def _on_calibration_done(self, future, roc_num):
+        """
+        Called when the calibrations for one region of calibration (roc) are finished
+        (either successfully, cancelled or failed).
+        It stores the calibrated parameters on the ROC object.
+        :param future: (ProgressiveFuture) Calibration future object, which can be cancelled.
+        :param roc_num: (int) The number of the region of calibration.
+        """
+
+        try:
+            config = future.result()  # wait until the calibration is done
+            self.save_calibrated_settings(roc_num, config)  # save calibrated
+        except CancelledError:
+            self.is_aligned.value = False  # don't enable ROA acquisition
+            self._update_calibration_controls("Calibration cancelled")  # update label to indicate cancelling
+            # TODO @Eric is then the callback for all futures still called?
+            # return
+        except Exception as ex:
+            logging.exception("Calibration failed with %s.", ex)
+            # return
+
+    @call_in_wx_main  # as changes in gui are triggered
+    def _on_batch_calibrations_done(self, batch_future):
+        """
+        Called when all calibrations are finished (either successfully, cancelled or failed).
+        :param batch_future: (ProgressiveFuture) Calibration future object, which can be cancelled.
+        """
+
+        self._tab_data.main.is_acquiring.value = False
+        self._future_connector = None  # reset connection to the progress bar
+
+        try:
+            batch_future.result()  # wait until the calibration is done
+            self.is_aligned.value = True  # allow acquiring ROAs
+            self._update_calibration_controls("Calibration successful")
+        except CancelledError:
+            self.is_aligned.value = False  # don't enable ROA acquisition
+            self._update_calibration_controls("Calibration cancelled")  # update label to indicate cancelling
+            # return
+        except Exception as ex:
+            self.is_aligned.value = False  # don't enable ROA acquisition
+            logging.exception("Calibration failed with %s.", ex)
+            self._update_calibration_controls("Calibration failed")
+            # return
```

Comment:
I think all these `# return`'s can be removed right?

Reply:
I am not sure. 
The cancelling is not raised correctly for some reason atm. 
And in the ROA acquisition and overview acquisition there is a return. 
@pieleric can you comment please?

Reply:
What happens at the moment is that when I cancel the calibration, it raises the CancellationError in the callback of the single future, but then in the batch future it is not receiving the CancellationError.